presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1049,41 +1054,68 @@ public PlanWithProperties visitIndexSource(IndexSourceNode node, Context context
             Optional<PreferredProperties.Global> parentGlobal = parentPreference.getGlobalProperties();
             if (parentGlobal.isPresent() && parentGlobal.get().isDistributed() && parentGlobal.get().getPartitioningProperties().isPresent()) {
-                PreferredProperties.PartitioningProperties partitioningPreference = parentGlobal.get().getPartitioningProperties().get();
-
-                // Use the requested hash partitioning, otherwise arbitrarily establish a hash partitioning column order
-                List<Symbol> hashingColumns = partitioningPreference.getHashingOrder().orElse(ImmutableList.copyOf(partitioningPreference.getPartitioningColumns()));
+                PreferredProperties.PartitioningProperties parentPartitioningPreference = parentGlobal.get().getPartitioningProperties().get();
+                Partitioning desiredParentPartitioning = selectUnionPartitioning(node, context, parentPartitioningPreference);
 
                 ImmutableList.Builder<PlanNode> partitionedSources = ImmutableList.builder();
                 ImmutableListMultimap.Builder<Symbol, Symbol> outputToSourcesMapping = ImmutableListMultimap.builder();
 
                 for (int sourceIndex = 0; sourceIndex < node.getSources().size(); sourceIndex++) {
-                    ImmutableList.Builder<Symbol> hashColumnsBuilder = ImmutableList.builder();
-                    for (Symbol column : hashingColumns) {
-                        hashColumnsBuilder.add(node.getSymbolMapping().get(column).get(sourceIndex));
-                    }
-                    List<Symbol> sourceHashColumns = hashColumnsBuilder.build();
+                    Partitioning childPartitioning = desiredParentPartitioning.translate(createTranslator(createMapping(node.getOutputSymbols(), node.sourceOutputLayout(sourceIndex))));
 
                     PreferredProperties childPreferred = PreferredProperties.builder()
-                            .global(PreferredProperties.Global.distributed(PreferredProperties.PartitioningProperties.hashPartitioned(sourceHashColumns)
-                                    .withNullsReplicated(partitioningPreference.isNullsReplicated())))
+                            .global(PreferredProperties.Global.distributed(PreferredProperties.PartitioningProperties.partitioned(childPartitioning)
+                                    .withNullsReplicated(parentPartitioningPreference.isNullsReplicated())))
                             .build();
+
                     PlanWithProperties source = node.getSources().get(sourceIndex).accept(this, context.withPreferredProperties(childPreferred));
-                    if (!source.getProperties().isNodePartitionedOn(Partitioning.create(FIXED_HASH_DISTRIBUTION, sourceHashColumns), partitioningPreference.isNullsReplicated())) {
+                    if (!source.getProperties().isNodePartitionedOn(childPartitioning, parentPartitioningPreference.isNullsReplicated())) {
```

Comment:
It is confusing to see `parentPartitioningPreference` here.  I think it would be better to use `desiredParentPartitioning` which will be the same anyway.